### ISSUE TITLE
Server can not bind to port 5683

### ISSUE THREAD

**Issue by lzheng5** [smeshlink/CoAP.NET#51]
Hi, 

Please see this [post](http://stackoverflow.com/questions/38468481/coap-net-can-not-bind-to-port-5683).

I am not sure about why it gives me that error. 

Thanks,

**Comment by longshine** [comment]
Hi~

It seems there might be something incorrect with the IPv6 network interface on your OS.

The log says:

```bash
DEBUG - Starting CoAP server
Warn - Cannot start endpoint at :::5683
DEBUG - Stopping endpoint bound to :::5683
Warn - Could not start endpoint :::5683
System.Net.Sockets.SocketException: The descriptor is not a socket
```

The server tries to bind a socket to an IPv6 address of _[::]_ (`System.Net.IPAddress.IPv6Any`) and fails.

Please try adding the following code **before** starting a server:

```csharp
server.AddEndPoint(IPAddress.Any, server.Config.DefaultPort);
```

It will bind to the IPv4 address (0.0.0.0), instead of the default IPv6 address.
You can use any other network address on your OS if you'd like to.

**Comment by lzheng5** [comment]
Hi, longshine

I have tried your suggestion. It still has problem with binding. 

Here is the error message.

```
Warn - Cannot start endpoint at 0.0.0.0:5683
DEBUG - Stopping endpoint bound to 0.0.0.0:5683
Warn - Could not start endpoint 0.0.0.0:5683
System.Net.Sockets.SocketException: The descriptor is not a socket
  at System.Net.Sockets.Socket.IOControl (Int32 ioctl_code, System.Byte[] in_value, System.Byte[] out_value) [0x00033] in /build/mono/src/mono-4.4.0/mcs/class/System/System.Net.Sockets/Socket.cs:3219 
  at CoAP.Channel.UDPChannel.SetupUDPSocket (AddressFamily addressFamily, Int32 bufferSize) [0x00009] in <filename unknown>:0 
  at CoAP.Channel.UDPChannel.Start () [0x00117] in <filename unknown>:0 
  at CoAP.Net.CoAPEndPoint.Start () [0x00040] in <filename unknown>:0 
```

It can not even bind to ipv4 localhost. 

The only distinction my server program has is that I have not added all the resources but HelloworldResource only. But I think this shall not affect the binding, right?

In order to rule out the possibility that my system has some problem, I have compiled a C program, and server can bind correctly.

Thanks,

**Comment by longshine** [comment]
Hi~

That's very odd. I googled and found some other projects have the same issue too, but no solution. See [rabbitmq-dotnet-client](https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/23) and [sockets-for-pcl](https://github.com/rdavisau/sockets-for-pcl/issues/69).

CoAP resources should have nothing to do with this issue.

I found some messages in the log:
```bash
  at System.Net.Sockets.Socket.IOControl (Int32 ioctl_code, System.Byte[] in_value, System.Byte[] out_value) [0x00033] in /build/mono/src/mono-4.4.0/mcs/class/System/System.Net.Sockets/Socket.cs:3219 
  at CoAP.Channel.UDPChannel.SetupUDPSocket (AddressFamily addressFamily, Int32 bufferSize) [0x00009] in <filename unknown>:0 
```

There are some preparation steps when setting up a UDP socket in `UDPChannel.SetupUDPSocket()`, which might not be well supported on Mono.
I will make a patch later. Are you using the binaries or the source codes?

**Comment by longshine** [comment]
Fixed in commit 7ffdb9420d282d6cd4c1abd0eea04b0995157ed9.

**Comment by lzheng5** [comment]
Hi, longshine

This bug has been resolved.

I can get the following result. 

```
DEBUG - Starting endpoint bound to 0.0.0.0:5683
CoAP server [RFC 7252] is listening on 0.0.0.0:5683
```

Even if `server.AddEndPoint(IPAddress.Any,server.Config.DefaultPort);` is commented out.

Thanks,
